rq/worker.py

```diff
@@ -667,24 +669,36 @@ def perform_job(self, job, queue):
                 # to use the same exc handling when pickling fails
                 job._result = rv
 
-                self.set_current_job_id(None, pipeline=pipeline)
-
                 result_ttl = job.get_result_ttl(self.default_result_ttl)
                 if result_ttl != 0:
                     job.ended_at = utcnow()
-                    job.set_status(JobStatus.FINISHED, pipeline=pipeline)
-                    job.save(pipeline=pipeline)
 
-                    finished_job_registry = FinishedJobRegistry(job.origin,
-                                                                self.connection)
-                    finished_job_registry.add(job, result_ttl, pipeline)
+                while True:
+                    try:
+                        # if dependencies are inserted after enqueue_dependents
+                        # a WatchError is thrown by execute()
+                        pipeline.watch(job.dependents_key)
```

Comment:
I think we can limit the scope of the `while` loop and `try/except WatchError` to `queue.enqueue_dependents`. That way, it will be much easier to understand since the only place that can raise `WatchError` is `queue.enqueue_dependents()`.

Could you also move the entire block of code that handles successful job execution to a separate method to make this method easier to digest? We already have `handle_job_failure` so I think we can call it `handle_successful_job_execution` (I'm terrible at naming so better suggestions are welcome :)).